### ISSUE TITLE
[BUG]: steps can't be configured with step config subclasses (for example slack alerter)

### ISSUE THREAD

**Issue by strangemonad** [zenml-io/zenml#700]
### Contact Details [Optional]

shawnmorel@gmail.com

### System Information

ZenML version: 0.8.1
Install path: /Users/shawnmorel/Code/rivian/me/factory-data/factory-data-algorithms/projects/common/process-health/.venv/lib/python3.9/site-packages/zenml
Python version: 3.9.9
Platform information: {'os': 'mac', 'mac_version': '12.4'}
Environment: native
Integrations: ['aws', 'kubeflow', 'scipy', 'seldon', 'slack']

### What happened?

When you try to use the `zenml.alerter.alerter_step` to configure it with a `SlackAlerterConfig`, the config arguments are silently lost.

**Cause**
- when preparing the tfx execution params the SlackAlerterConfig is **correctly** serialized to json (see the pipeline protobuf in "relevant log output below")
- however when inferring the `alerter_step` input signature, the spec.args type is set to `BaseAlerterStepConfig` (the type annotation of the `alerter_step` function. In reality this function can take any `T <: BaseAlerterStepConfig` (BaseAlerterStepConfig is a type bound). One way to capture this would be to override the actual arg type at the call site `alerter_step(config=SlackAlerterConfig)` asserting that it is an instance of `BaseAlerterStepConfig` but storing the concrete config subclass type
- when the tfx component launcher tries to run the step, it deserializes execution params = `{slack_channel_id="ABC123"}` into a BaseAlerterStepConfig -> `steps.util.py:426 `config_objject = arg_type.parse_obj(exec_properties)`. Pydantic silently ignores the `slack_channel_id` field and returns an empty BaseAlerterConfig.

**What I'd expect to happen**
1. Don't silently fail to deserialize the params on execution, I'd much rather see something like `ValueError: failed to create 'BaseAlerterStepConfig' with arguments '{slack_channel_id="ABC123"}'`. This probably means (at least for arg specs) pydantic should be configured to not allow unknown fields whenever the concrete subtype can be inferred from the call site.
2. Change the signature of alerter_step to be generic and have `BaseAlerterStepConfig` as an upper bound.
3. Possibly change the behavior of how input param type signatures get captured to capture the concrete sub-type a call time and de-serialize using that concrete sub-type during execution.

In general, I know the API for alerters is still being worked on and this is a good design case to keep working out interfaces where different concrete stack components behave differently. Given that a pipeline should ideally be agnostic of the stack it's running on, I think it's important to keep figuring out how to keep the pipeline definition cleanly separated from the connecting and running :hands

```
@step
def msg() -> str:
    return "hello slack"

@pipeline
def slack_pipeline(msg, alerter):
    m = msg()
    alerter(m)

def main():
    slack_pipeline(
        msg=msg(),
        alerter=alerter_step(SlackAlerterConfig(
            slack_channel_id="ABC123",
        )),
    ).run()
```

### Reproduction steps

Run the pipeline described above with the following stack configuration

```
        Stack Configuration
┏━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━┓
┃ COMPONENT_TYPE │ COMPONENT_NAME ┃
┠────────────────┼────────────────┨
┃ ALERTER        │ slack_alerter  ┃
┠────────────────┼────────────────┨
┃ ARTIFACT_STORE │ default        ┃
┠────────────────┼────────────────┨
┃ METADATA_STORE │ default        ┃
┠────────────────┼────────────────┨
┃ ORCHESTRATOR   │ default
````

### Relevant log output

```shell
pipeline_info {
  id: "slack_pipeline"
}
nodes {
  pipeline_node {
    node_info {
      type {
        name: "__main__.msg"
      }
      id: "msg"
    }
    contexts {
      contexts {
        type {
          name: "pipeline"
        }
        name {
          field_value {
            string_value: "slack_pipeline"
          }
        }
      }
      contexts {
        type {
          name: "pipeline_run"
        }
        name {
          runtime_parameter {
            name: "pipeline-run-id"
            type: STRING
          }
        }
      }
      contexts {
        type {
          name: "node"
        }
        name {
          field_value {
            string_value: "slack_pipeline.msg"
          }
        }
      }
    }
    outputs {
      outputs {
        key: "output"
        value {
          artifact_spec {
            type {
              name: "DataArtifact"
              properties {
                key: "datatype"
                value: STRING
              }
              properties {
                key: "materializer"
                value: STRING
              }
            }
          }
        }
      }
    }
    parameters {
      parameters {
        key: "zenml-custom_step_operator"
        value {
          field_value {
            string_value: "null"
          }
        }
      }
      parameters {
        key: "zenml-output_materializer_source"
        value {
          field_value {
            string_value: "\"ff47c7d1394cef1d90d1169fcade0ba9448d3d87fc5fa7c0fe0df56846674823\""
          }
        }
      }
      parameters {
        key: "zenml-pipeline_parameter_name"
        value {
          field_value {
            string_value: "\"msg\""
          }
        }
      }
      parameters {
        key: "zenml-step_source"
        value {
          field_value {
            string_value: "\"54d3616b74a8aaf48169aba3fa134f8cd80b30f1e587539b6c539f7c7dcdd620\""
          }
        }
      }
    }
    downstream_nodes: "alerter_step"
    execution_options {
      caching_options {
        enable_cache: true
      }
    }
  }
}
nodes {
  pipeline_node {
    node_info {
      type {
        name: "zenml.alerter.alerter_step.alerter_step"
      }
      id: "alerter_step"
    }
    contexts {
      contexts {
        type {
          name: "pipeline"
        }
        name {
          field_value {
            string_value: "slack_pipeline"
          }
        }
      }
      contexts {
        type {
          name: "pipeline_run"
        }
        name {
          runtime_parameter {
            name: "pipeline-run-id"
            type: STRING
          }
        }
      }
      contexts {
        type {
          name: "node"
        }
        name {
          field_value {
            string_value: "slack_pipeline.alerter_step"
          }
        }
      }
    }
    inputs {
      inputs {
        key: "message"
        value {
          channels {
            producer_node_query {
              id: "msg"
            }
            context_queries {
              type {
                name: "pipeline"
              }
              name {
                field_value {
                  string_value: "slack_pipeline"
                }
              }
            }
            context_queries {
              type {
                name: "pipeline_run"
              }
              name {
                runtime_parameter {
                  name: "pipeline-run-id"
                  type: STRING
                }
              }
            }
            context_queries {
              type {
                name: "node"
              }
              name {
                field_value {
                  string_value: "slack_pipeline.msg"
                }
              }
            }
            artifact_query {
              type {
                name: "DataArtifact"
              }
            }
            output_key: "output"
          }
          min_count: 1
        }
      }
    }
    outputs {
      outputs {
        key: "output"
        value {
          artifact_spec {
            type {
              name: "DataArtifact"
              properties {
                key: "datatype"
                value: STRING
              }
              properties {
                key: "materializer"
                value: STRING
              }
            }
          }
        }
      }
    }
    parameters {
      parameters {
        key: "slack_channel_id"
        value {
          field_value {
            string_value: "\"ABC123\""
          }
        }
      }
      parameters {
        key: "zenml-custom_step_operator"
        value {
          field_value {
            string_value: "null"
          }
        }
      }
      parameters {
        key: "zenml-disable_cache"
        value {
          field_value {
            string_value: "\"dfae913e3230107f3b6cd1c7c6fc0402\""
          }
        }
      }
      parameters {
        key: "zenml-pipeline_parameter_name"
        value {
          field_value {
            string_value: "\"alerter\""
          }
        }
      }
    }
    upstream_nodes: "msg"
    execution_options {
      caching_options {
        enable_cache: true
      }
    }
  }
}
runtime_spec {
  pipeline_root {
    runtime_parameter {
      name: "pipeline-root"
      type: STRING
      default_value {
        string_value: "/Users/shawnmorel/Library/Application Support/zenml/local_stores/00df0bed-a563-495d-8217-db08822a449b"
      }
    }
  }
  pipeline_run_id {
    runtime_parameter {
      name: "pipeline-run-id"
      type: STRING
    }
  }
}
execution_mode: SYNC
deployment_config {
  type_url: "type.googleapis.com/tfx.orchestration.IntermediateDeploymentConfig"
  value: "\n\226\001\n\014alerter_step\022\205\001\nOtype.googleapis.com/tfx.orchestration.executable_spec.PythonClassExecutableSpec\0222\n0zenml.alerter.alerter_step.alerter_step_Executor\nq\n\003msg\022j\nOtype.googleapis.com/tfx.orchestration.executable_spec.PythonClassExecutableSpec\022\027\n\025__main__.msg_Executor*\253\001\n0type.googleapis.com/ml_metadata.ConnectionConfig\022w\032u\nq/Users/shawnmorel/Library/Application Support/zenml/local_stores/00df0bed-a563-495d-8217-db08822a449b/metadata.db\020\003"
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by strangemonad** [comment]
Also, just to confirm that this really is a serializer issue, here's my current work-around that is able to send messages without any issues (it's a slightly modified version of `alerter_step` that takes a SlackAlerterConfig in it's method signature

```
from zenml.exceptions import DoesNotExistException
from zenml.integrations.slack.alerters.slack_alerter import SlackAlerterConfig, SlackAlerter
from zenml.steps import step, StepContext

logger = logging.get_logger()


@step
def slack_alerter(
    config: SlackAlerterConfig,
    context: StepContext,
    message: str,
) -> bool:
    """Custom slack_alerter step as a temporary workaround for a zenml bug.

    https://github.com/zenml-io/zenml/issues/700

    Args
    ----
        config: The slack alerter configuration
        context: StepContext of the ZenML repository.
        message: Message to be posted.

    Returns
    -------
        True if operation succeeded, else False.
    """

    if not context.stack:
        raise DoesNotExistException(
            "No active stack is available. "
            "Please make sure that you have registered and set a stack."
        )
    if not context.stack.alerter:
        raise ValueError(
            "The active stack don't have any alerter configured",
            context.stack,
        )

    if not isinstance(context.stack.alerter, SlackAlerter):
        raise ValueError(
            "The current stack's alerter is not a slack alerter",
            context.stack.alerter,
        )

    return context.stack.alerter.post(message, config)
````

**Comment by fa9r** [comment]
Hi Shawn, thanks so much for the issue. 

I'm really amazed by your analysis, that is exactly what is happening indeed. We were discussing this issue internally and are considering removing the general `alerter_step` in favor of an integration-specific `slack_alerter_step`. However, we are still not 100% decided, as integration-specific steps mean that you need to change your code when switching stacks, which we generally want to avoid.

For now, using the custom `slack_alerter` step you posted above is the best way to configure your alerter at runtime. Again, big kudos and thanks for posting that 👏 

Best regards,
Felix

**Comment by strangemonad** [comment]
@fa9r no worries. Thanks for an amazing framework!

For what it's worth here's what I would like to see.

1. I think the option to have specific alerters makes a ton of sense. By and large, internally we're always going to use the same chat ops and the stacks are really more about moving from DEV slack token / channel to PROD slack token and channel. A specific slack alerter step allows deeper integration with the specific APIs e.g. I can post a fully featured slack card that unfurls or other rich content. In this world, though, I'd still want a default "slack alerter" that simply uses the common logging api when running the default stack locally or in CI unit tests and integrations tests.

2. I think I'd still like a generic alerter. Not having a generic alerter means that if ever internally we want to move from slack to something else, it would be a massive effort to test compatibility. If you don't have a generic alerter, it means you also prevent a marketplace of re-usable step and pipeline components. You could still allow some rich content (e.g. allow registering different renderers for images, markdown, html content). I think a similar solution could be applied to step visualizers (e.g. kill all the `if jupyter` statements and instead, the context knows if you're running in KFP or jupyter). In my mind, this is exactly like a SQL abstraction layer where by default you code to a common API and in specific cases you check if you're running on postgres vs mysql in case you need to use a db-specific operation.



In general, my mind keeps going to guice or dagger style dependency injection if you're familiar with that. The stacks and pipeline config aren't 100% cutting things at the right abstraction level for me it would seem. I definitely agree with the philosophy that pipelines should be generic and that the stack should provide the runtime binding. As written though, config muddies the water between pipeline configuration (e.g. per environment I want to use a different slack channel, the data scientist doesn't care, this is something the MLOps engineer sets up for DEV, STAGE, PROD) and pipeline params that semantically affect my pipeline operation (e.g. model hyper-params).

**Comment by fa9r** [comment]
@strangemonad @amirhessam88 I temporarily fixed this issue in #714 by removing the general alerter steps and replacing them by Slack-specific steps. This will be live after the next release (probably end of next week). @strangemonad thanks again for bringing up the issue!
